src/magic_folder/test/test_local_snapshot.py

```diff
@@ -0,0 +1,336 @@
+    Always,
+    HasLength,
+    MatchesStructure,
+)
+from testtools.twistedsupport import (
+    succeeded,
+)
+from testtools import (
+    ExpectedException,
+)
+
+from eliot import (
+    Message,
+)
+
+from magic_folder.magic_folder import (
+    LocalSnapshotService,
+    LocalSnapshotCreator,
+)
+from magic_folder.snapshot import (
+    create_local_author,
+)
+from magic_folder.magicpath import (
+    path2magic,
+)
+from .. import magicfolderdb
+
+from .common import (
+    SyncTestCase,
+)
+from .strategies import (
+    path_segments,
+)
+
+@attr.s
+class MemorySnapshotCreator(object):
+    """
+    A way to test LocalSnapshotService with an in-memory database.
+    """
+    processed = attr.ib(default=attr.Factory(dict))
+    def process_item(self, path):
+        Message.log(
+            message_type=u"memory-snapshot-creator:process_item",
+            path=path.asTextMode("utf-8").path,
+        )
+        if path not in self.processed.keys():
+            # record number of parents
+            self.processed[path] = 0
+        else:
+            num_parents = self.processed[path]
+            self.processed[path] = num_parents + 1
+        return defer.succeed(None)
+
+
+class LocalSnapshotServiceTests(SyncTestCase):
+
+    def setUp(self):
+        super(LocalSnapshotServiceTests, self).setUp()
+        self.db = magicfolderdb.get_magicfolderdb(":memory:", create_version=(magicfolderdb.SCHEMA_v1, 1))
+        self.author = create_local_author("alice")
+
+        self.stash_dir = FilePath(self.mktemp())
+        self.stash_dir.makedirs()
+
+        self.magic_path = FilePath(self.mktemp())
+        self.magic_path.makedirs()
+
+        self._snapshot_creator = MemorySnapshotCreator()
+
+        self.snapshot_service = LocalSnapshotService(
+            magic_path=self.magic_path,
+            snapshot_creator=self._snapshot_creator,
+        )
+
+    def setup_example(self):
+        """
+        Hypothesis-invoked hook to create per-example state.
+        Reset the database before running each test.
+        """
+        self._snapshot_creator.processed = {}
+
+    def test_add_single_file(self):
+        """
+        Start the service, add a file and check if the operation succeeded.
+        """
+        foo = self.magic_path.child("foo")
+        content = u"foo"
+        with foo.open("w") as f:
+            f.write(content)
+
+        self.snapshot_service.startService()
+        d = self.snapshot_service.add_file(foo)
+
+        self.assertThat(
+            d,
+            succeeded(Always()),
+        )
+
+        self.assertThat(
+            self.snapshot_service.stopService(),
+            succeeded(Always())
+        )
+
+        self.assertThat(self._snapshot_creator.processed.get(foo), Not(Equals(None)))
+        self.assertThat(self._snapshot_creator.processed.get(foo), Equals(0))
+
+    @given(lists(path_segments().map(lambda p: p.encode("utf-8")), unique=True),
+           lists(binary(), unique=True))
```

Comment:
Probably doesn't matter if the contents are unique?

What does matter is the the lists are the same length - which they won't currently always be since the separate `lists` strategies are free to pick their own lengths.  The solution here is probably to use the `data` strategy to draw examples imperatively inside the body of the test.

```
@given(lists(...), data()):
def test_add_multiple_files(self, filenames, data):
    for filename in filenames:
        content = data.draw(binary())
        ...
```

`data.draw(binary())` will return an example like the one that would have been passed to the function if you'd passed `binary()` to the `given` decorator.

Reply:
Thanks. I didn't know about data() strategy.